File: examples/CXX/hints.cpp
/*********************************************************************
 *
 *  Copyright (C) 2014, Northwestern University and Argonne National Laboratory
 *  See COPYRIGHT notice in top-level directory.
 *
 *********************************************************************/
/* $Id$ */

/* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
 * This example sets two PnetCDF hints:
 *    nc_header_align_size and nc_var_align_size
 * and prints the hint values as well as the header size, header extent, and
 * two variables' starting file offsets.
 *
 * The compile and run commands are given below.
 *
 *    % mpicxx -O2 -o hints hints.c -lpnetcdf
 *
 *    % mpiexec -l -n 4 ./hints /pvfs2/wkliao/testfile.nc
 *
 *    nc_header_align_size      set to = 1024
 *    nc_var_align_size         set to = 512
 *    nc_header_read_chunk_size set to = 256
 *    header size                      = 252
 *    header extent                    = 1024
 *    var_zy start file offset         = 1024
 *    var_yx start file offset         = 3072
 *
 * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */

#include <stdio.h>
#include <stdlib.h>

#include <iostream>
using namespace std;

#include <string.h> /* strcpy(), strncpy() */
#include <unistd.h> /* getopt() */
#include <pnetcdf>

using namespace PnetCDF;
using namespace PnetCDF::exceptions;

#define NZ 5
#define NY 5
#define NX 5

static void
usage(char *argv0)
{
    cerr <<
    "Usage: %s [-h] | [-q] [file_name]\n"
    "       [-h] Print help\n"
    "       [-q] Quiet mode (reports when fail)\n"
    "       [filename] output netCDF file name\n"
    << argv0;
}

static
void print_hints(NcmpiFile &ncFile,
                 NcmpiVar  &var0,
                 NcmpiVar  &var1)
{
    char value[MPI_MAX_INFO_VAL];
    int len, flag;
    MPI_Offset header_size, header_extent, var_zy_start, var_yx_start;
    MPI_Offset h_align=-1, v_align=-1, h_chunk=-1;
    MPI_Info info_used;

    ncFile.Inq_header_size(&header_size);
    ncFile.Inq_header_extent(&header_extent);

    var0.Inq_file_offset(&var_zy_start);
    var1.Inq_file_offset(&var_yx_start);

    /* get all the hints used */
    ncFile.Inq_file_info(&info_used);

    MPI_Info_get_valuelen(info_used, (char*)"nc_header_align_size", &len, &flag);
    if (flag) {
        MPI_Info_get(info_used, (char*)"nc_header_align_size", len+1, value, &flag);
        h_align = strtoll(value,NULL,10);
    }
        MPI_Info_get_valuelen(info_used, (char*)"nc_var_align_size", &len, &flag);
    if (flag) {
        MPI_Info_get(info_used, (char*)"nc_var_align_size", len+1, value, &flag);
        v_align = strtoll(value,NULL,10);
    }
    MPI_Info_get_valuelen(info_used, (char*)"nc_header_read_chunk_size", &len, &flag);
    if (flag) {
        MPI_Info_get(info_used, (char*)"nc_header_read_chunk_size", len+1, value,&flag);
        h_chunk = strtoll(value,NULL,10);
    }
    MPI_Info_free(&info_used);

    if (h_align == -1)
        printf("nc_header_align_size      is NOT set\n");
    else
        printf("nc_header_align_size      set to = %lld\n", h_align);

    if (v_align == -1)
        printf("nc_var_align_size         is NOT set\n");
    else
        printf("nc_var_align_size         set to = %lld\n", v_align);
    if (h_chunk == -1)
        printf("nc_header_read_chunk_size is NOT set\n");
    else
        printf("nc_header_read_chunk_size set to = %lld\n", h_chunk);

    printf("header size                      = %lld\n", header_size);
    printf("header extent                    = %lld\n", header_extent);
    printf("var_zy start file offset         = %lld\n", var_zy_start);
    printf("var_yx start file offset         = %lld\n", var_yx_start);
}

int main(int argc, char** argv)
{
    extern int optind;
    char filename[256];
    int i, rank, nprocs, verbose=1, *buf_zy;
    float *buf_yx;
    vector<MPI_Offset> start(2), count(2);
    MPI_Info info;

    MPI_Init(&argc, &argv);
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    MPI_Comm_size(MPI_COMM_WORLD, &nprocs);

    /* get command-line arguments */
    while ((i = getopt(argc, argv, "hq")) != EOF)
        switch(i) {
            case 'q': verbose = 0;
                      break;
            case 'h':
            default:  if (rank==0) usage(argv[0]);
                      MPI_Finalize();
                      return 1;
        }
    if (argv[optind] == NULL) strcpy(filename, "testfile.nc");
    else                      snprintf(filename, 256, "%s", argv[optind]);

    try {
        MPI_Info_create(&info);
        MPI_Info_set(info, (char*)"nc_header_align_size",      (char*)"1024");
        MPI_Info_set(info, (char*)"nc_var_align_size",         (char*)"512");
        MPI_Info_set(info, (char*)"nc_header_read_chunk_size", (char*)"256");
        /* note that set the above values to 1 to disable the alignment */

        /* create a new file for writing -------------------------------------*/
        NcmpiFile ncFile(MPI_COMM_WORLD, filename, NcmpiFile::replace,
                         NcmpiFile::classic5, info);
        MPI_Info_free(&info);

        /* define 3 dimensions */
        vector<NcmpiDim> dimid(3);
        dimid[0] = ncFile.addDim("Z", NZ*nprocs);
        dimid[1] = ncFile.addDim("Y", NY*nprocs);
        dimid[2] = ncFile.addDim("X", NX*nprocs);

        /* define a variable of size (NZ * nprocs) * (NY * nprocs) */
        vector<NcmpiDim> dimid_zy(2);
        dimid_zy[0] = dimid[0];
        dimid_zy[1] = dimid[1];
        NcmpiVar var0 = ncFile.addVar("var_zy", ncmpiInt, dimid_zy);

        /* define a variable of size (NY * nprocs) * (NX * nprocs) */
        vector<NcmpiDim> dimid_yx(2);
        dimid_yx[0] = dimid[1];
        dimid_yx[1] = dimid[2];
        NcmpiVar var1 = ncFile.addVar("var_yx", ncmpiFloat, dimid_yx);

        /* var_zy is partitioned along Z dimension */
        buf_zy = (int*) malloc(NZ * (NY * nprocs) * sizeof(int));
        for (i=0; i<NZ*(NY*nprocs); i++) buf_zy[i] = i;

        start[0] = NZ * rank; start[1] = 0;
        count[0] = NZ;        count[1] = NY * nprocs;
        var0.putVar_all(start, count, buf_zy);

        /* var_yx is partitioned along X dimension */
        buf_yx = (float*) malloc((NY * nprocs) * NX * sizeof(float));
        for (i=0; i<(NY*nprocs)*NX; i++) buf_yx[i] = i;

        start[0] = 0;           start[1] = NX * rank;
        count[0] = NY * nprocs; count[1] = NX;
        var1.putVar_all(start, count, buf_yx);

        if (rank == 0 && verbose) print_hints(ncFile, var0, var1);

        free(buf_zy);
        free(buf_yx);
    }
    catch(NcmpiException& e) {
       cout << e.what() << " error code=" << e.errorCode() << " Error!\n";
       return 1;
    }

    /* check if there is any PnetCDF internal malloc residue */
    MPI_Offset malloc_size, sum_size;
    int err = ncmpi_inq_malloc_size(&malloc_size);
    if (err == NC_NOERR) {
        MPI_Reduce(&malloc_size, &sum_size, 1, MPI_OFFSET, MPI_SUM, 0, MPI_COMM_WORLD);
        if (rank == 0 && sum_size > 0)
            printf("heap memory allocated by PnetCDF internally has %lld bytes yet to be freed\n",
                   sum_size);
    }

    MPI_Finalize();
    return 0;
}

File: examples/F90/hints.f90
!
!
!  Copyright (C) 2012, Northwestern University and Argonne National
!  Laboratory
!  See COPYRIGHT notice in top-level directory.
!
! $Id$

!
! This example sets two PnetCDF hints:
!    nc_header_align_size and nc_var_align_size
! and prints the hint values as well as the header size, header extent, and
! two variables' starting file offsets.
!
! The compile and run commands are given below.
!
!    % mpif90 -O2 -o hints hints.f90 -lpnetcdf
!
!    % mpiexec -n 4 ./hints /pvfs2/wkliao/testfile.nc
!
!    nc_header_align_size      set to = 1024
!    nc_var_align_size         set to = 512
!    nc_header_read_chunk_size set to = 256
!    header size                      = 252
!    header extent                    = 1024
!    var_zy start file offset         = 1024
!    var_yx start file offset         = 3072
!


      subroutine check(err, message)
          use mpi
          use pnetcdf
          implicit none
          integer err
          character(len=*) message

          ! It is a good idea to check returned value for possible error
          if (err .NE. NF90_NOERR) then
              write(6,*) trim(message), trim(nf90mpi_strerror(err))
              call MPI_Abort(MPI_COMM_WORLD, -1, err)
          end if
      end subroutine check

      subroutine print_hints(ncid, varid0, varid1)
          use mpi
          use pnetcdf
          implicit none

          integer ncid, varid0, varid1

          ! local variables
          character*(MPI_MAX_INFO_VAL) value
          integer err, len, info_used
          logical flag
          integer(kind=MPI_OFFSET_KIND) header_size, header_extent
          integer(kind=MPI_OFFSET_KIND) var_zy_start, var_yx_start
          integer(kind=MPI_OFFSET_KIND) h_align, v_align, h_chunk

          h_align=-1
          v_align=-1
          h_chunk=-1

          err = nf90mpi_inq_header_size  (ncid, header_size)
          call check(err, 'In nf90mpi_inq_header_size: ')
          err = nf90mpi_inq_header_extent(ncid, header_extent)
          call check(err, 'In nf90mpi_inq_header_extent: ')
          err = nf90mpi_inq_varoffset(ncid, varid0, var_zy_start)
          call check(err, 'In nf90mpi_inq_varoffset varid0: ')
          err = nf90mpi_inq_varoffset(ncid, varid1, var_yx_start)
          call check(err, 'In nf90mpi_inq_varoffset varid1: ')

          err = nf90mpi_inq_file_info(ncid, info_used)
          call check(err, 'In nf90mpi_inq_file_info : ')

          call MPI_Info_get_valuelen(info_used, "nc_header_align_size", &
                                     len, flag, err)
          if (flag) then
              call MPI_Info_get(info_used, "nc_header_align_size", &
                                len+1, value, flag, err)
              read(value, '(i16)') h_align
          endif
          call MPI_Info_get_valuelen(info_used, "nc_var_align_size", &
                                     len, flag, err)
          if (flag) then
              call MPI_Info_get(info_used, "nc_var_align_size", len+1, &
                                value, flag, err)
              read(value, '(i16)') v_align
          endif
          call MPI_Info_get_valuelen(info_used, &
                                     "nc_header_read_chunk_size", &
                                     len, flag, err)
          if (flag) then
              call MPI_Info_get(info_used, "nc_header_read_chunk_size", &
                                len+1, value, flag, err)
              read(value, '(i16)') h_chunk
          endif
          call MPI_Info_free(info_used, err)

          if (h_align .EQ. -1) then
              print*,"nc_header_align_size      is NOT set"
          else
              print*,"nc_header_align_size      set to = ", h_align
          endif
          if (v_align .EQ. -1) then
              print*,"nc_var_align_size         is NOT set"
          else
              print*,"nc_var_align_size         set to = ", v_align
          endif
          if (h_chunk .EQ. -1) then
              print*,"nc_header_read_chunk_size is NOT set"
          else
              print*,"nc_header_read_chunk_size set to = ", h_chunk
          endif

          print*,"header size                      = ", header_size
          print*,"header extent                    = ", header_extent
          print*,"var_zy start file offset         = ", var_zy_start
          print*,"var_yx start file offset         = ", var_yx_start
      end subroutine print_hints

      subroutine put_vara(ncid, varid, ny, nx, start, count)
          use mpi
          use pnetcdf
          implicit none

          integer ncid, varid, ny, nx
          integer(KIND=MPI_OFFSET_KIND) start(2), count(2)

          integer i, j, rank, err
          integer buf(nx, ny)

          call MPI_Comm_rank(MPI_COMM_WORLD, rank, err)

          do j=1, ny
             do i=1, nx
                buf(i,j) = rank ! j*nx + i
             enddo
          enddo

          err = nf90mpi_put_var_all(ncid, varid, buf, start, count)
          call check(err, 'In nf90mpi_put_var_all : ')

      end subroutine put_vara

      program main
      use mpi
      use pnetcdf
      implicit none

      character(len = 256) :: filename, cmd
      integer NZ, NY, NX
      integer ncid, rank, nprocs, info, cmode, err, ierr, get_args, dummy
      integer varid0, varid1, dimid(3), dimid2(2)
      integer(KIND=MPI_OFFSET_KIND) start(2), count(2), llen
      integer(kind=MPI_OFFSET_KIND) malloc_size, sum_size
      logical verbose

      call MPI_Init(err)
      call MPI_Comm_rank(MPI_COMM_WORLD, rank, err)
      call MPI_Comm_size(MPI_COMM_WORLD, nprocs, err)

      NZ = 5
      NY = 5
      NX = 5

      if (rank .EQ. 0) then
          verbose = .TRUE.
          filename = "testfile.nc"
          ierr = get_args(2, cmd, filename, verbose, dummy)
      endif
      call MPI_Bcast(ierr, 1, MPI_INTEGER, 0, MPI_COMM_WORLD, err)
      if (ierr .EQ. 0) goto 999

      call MPI_Bcast(verbose, 1, MPI_LOGICAL, 0, MPI_COMM_WORLD, err)
      call MPI_Bcast(filename, 256, MPI_CHARACTER, 0, MPI_COMM_WORLD, err)

      call MPI_Info_create(info, err)
      call MPI_Info_set(info, "nc_header_align_size",      "1024", err)
      call MPI_Info_set(info, "nc_var_align_size",         "512",  err)
      call MPI_Info_set(info, "nc_header_read_chunk_size", "256",  err)
      ! note that set the above values to 1 to disable the alignment

      ! create a new file for writing
      cmode = IOR(NF90_CLOBBER, NF90_64BIT_DATA)
      err = nf90mpi_create(MPI_COMM_WORLD, filename, cmode, info, ncid)
      call check(err, 'In nf90mpi_create : ')
      call MPI_Info_free(info, err)

      ! define 3 dimensions
      llen = NZ*nprocs
      err = nf90mpi_def_dim(ncid, "Z", llen, dimid(1))
      call check(err, 'In nf90mpi_def_dim Z : ')
      llen = NY*nprocs
      err = nf90mpi_def_dim(ncid, "Y", llen, dimid(2))
      call check(err, 'In nf90mpi_def_dim Y : ')
      llen = NX*nprocs
      err = nf90mpi_def_dim(ncid, "X", llen, dimid(3))
      call check(err, 'In nf90mpi_def_dim X : ')

      ! define a variable of size (NZ * nprocs) * (NY * nprocs)
      dimid2(1) = dimid(1)
      dimid2(2) = dimid(2)
      err = nf90mpi_def_var(ncid, "var_zy", NF90_INT,   dimid2, varid0)
      call check(err, 'In nf90mpi_def_var var_zy : ')
      ! define a variable of size (NY * nprocs) * (NX * nprocs)
      dimid2(1) = dimid(2)
      dimid2(2) = dimid(3)
      err = nf90mpi_def_var(ncid, "var_yx", NF90_FLOAT, dimid2, varid1)
      call check(err, 'In nf90mpi_def_var var_yx : ')

      err = nf90mpi_enddef(ncid)
      call check(err, 'In nf90mpi_enddef : ')

      ! var_zy is partitioned along Z dimension
      start(1) = NZ * rank + 1
      start(2) = 1
      count(1) = NZ
      count(2) = NY * nprocs
      call put_vara(ncid, varid0, NZ, NY * nprocs, start, count)

      ! var_yx is partitioned along X dimension
      start(1) = 1
      start(2) = NX * rank + 1
      count(1) = NY * nprocs
      count(2) = NX
      call put_vara(ncid, varid1, NY * nprocs, NX, start, count)

      if (rank .EQ. 0 .AND. verbose) call print_hints(ncid, varid0, varid1)

      err = nf90mpi_close(ncid)

      ! check if there is any PnetCDF internal malloc residue
 998  format(A,I13,A)
      err = nf90mpi_inq_malloc_size(malloc_size)
      if (err == NF90_NOERR) then
          call MPI_Reduce(malloc_size, sum_size, 1, MPI_INTEGER8, &
                          MPI_SUM, 0, MPI_COMM_WORLD, err)
          if (rank .EQ. 0 .AND. sum_size .GT. 0_MPI_OFFSET_KIND) print 998, &
              'heap memory allocated by PnetCDF internally has ',  &
              sum_size/1048576, ' MiB yet to be freed'
      endif

 999  call MPI_Finalize(err)
      ! call EXIT(0) ! EXIT() is a GNU extension

      end program

